gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+   * Otherwise it should be set to true.
+   */
+  private void runCompactionForDataset(Dataset dataset, boolean proceed) {
+    LOG.info("Running compaction for dataset " + dataset);
+    MRCompactorJobRunner jobRunner = getMRCompactorJobRunner(dataset, dataset.priority());
+    this.jobRunnables.put(dataset, jobRunner);
+    if (proceed) {
+      jobRunner.proceed();
+    }
+    this.jobExecutor.execute(jobRunner);
   }
 
   /**
-   * Get an instance of MRCompactorJobRunner, e.g., MRCompactorAvroKeyJobRunner if the data is in Avro format.
+   * Get an instance of {@link MRCompactorJobRunner}.
    */
-  private MRCompactorJobRunner getMRCompactorJobRunner(State jobProps) {
+  private MRCompactorJobRunner getMRCompactorJobRunner(Dataset dataset, double priority) {
     try {
       @SuppressWarnings("unchecked")
       Class<? extends MRCompactorJobRunner> cls =
-          (Class<? extends MRCompactorJobRunner>) Class.forName(jobProps.getProp(
+          (Class<? extends MRCompactorJobRunner>) Class.forName(this.state.getProp(
               ConfigurationKeys.COMPACTION_JOB_RUNNER_CLASS, ConfigurationKeys.DEFAULT_COMPACTION_JOB_RUNNER_CLASS));
-      return cls.getDeclaredConstructor(State.class, FileSystem.class).newInstance(jobProps, this.fs);
+      return cls.getDeclaredConstructor(Dataset.class, FileSystem.class, Double.class).newInstance(dataset, this.fs,
+          priority);
     } catch (Exception e) {
       throw new RuntimeException("Cannot instantiate MRCompactorJobRunner", e);
     }
   }
 
-  public static void addRunningHadoopJob(Job job) {
-    MRCompactor.RUNNING_MR_JOBS.add(job);
+  /**
+   * Keep track of running MR jobs, so if the compaction is cancelled, the MR jobs can be killed.
+   */
+  public static void addRunningHadoopJob(Dataset dataset, Job job) {
+    MRCompactor.RUNNING_MR_JOBS.put(dataset, job);
   }
 
-  private long getMRJobTimeoutValue() {
+  private long getCompactionTimeoutMinutes() {
     return this.state.getPropAsLong(ConfigurationKeys.COMPACTION_MR_JOB_TIMEOUT_MINUTES,
         ConfigurationKeys.DEFAULT_COMPACTION_MR_JOB_TIMEOUT_MINUTES);
   }
 
+  private long getDataVerifTimeoutMinutes() {
+    return this.state.getPropAsLong(ConfigurationKeys.COMPACTION_DATA_VERIFICATION_TIMEOUT_MINUTES,
+        ConfigurationKeys.DEFAULT_COMPACTION_DATA_VERIFICATION_TIMEOUT_MINUTES);
+  }
+
+  private void throwExceptionsIfAnyDatasetCompactionFailed() {
+    Set<Dataset> datasetsWithThrowables = getDatasetsWithThrowables();
+    int numDatasetsWithThrowables = 0;
+    for (Dataset dataset : datasetsWithThrowables) {
+      numDatasetsWithThrowables++;
+      for (Throwable t : dataset.throwables()) {
+        LOG.error("Error processing dataset " + dataset, t);
+        CompactionSlaEventHelper.populateState(dataset.jobProps(), Optional.<Job> absent(), this.fs);
+        new SlaEventSubmitter(eventSubmitter, "CompactionFailed", dataset.jobProps().getProperties()).submit();
```

Comment:
Actually, reviewing more in depth the rest of the code, I think we need the event emission to happen on certain dataset state transitions, like like VERIFIED, GIVEN_UP, etc.